tools/const_finder/lib/const_finder.dart

```diff
@@ -0,0 +1,122 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:kernel/kernel.dart' hide MapEntry;
+import 'package:meta/meta.dart';
+
+class _ConstFinder  extends RecursiveVisitor<void> {
+  _ConstFinder(
+    this.kernelFilePath,
+    this.targetLibraryUri,
+    this.classLibraryUri,
+    this.className,
+  )  : assert(kernelFilePath != null),
+        assert(targetLibraryUri != null),
+        assert(classLibraryUri != null),
+        assert(className != null),
+        constantInstances = <Map<String, dynamic>>[],
+        nonConstantLocations = <Map<String, dynamic>>[];
+
+  /// The path to the file to open.
+  final String kernelFilePath;
+
+  /// The library URI for the main entrypoint of the target library.
+  final String targetLibraryUri;
+
+  /// The library URI for the class to find.
+  final String classLibraryUri;
+
+  /// The name of the class to find.
+  final String className;
+
+  final List<Map<String, dynamic>> constantInstances;
+  final List<Map<String, dynamic>> nonConstantLocations;
+
+  bool _matches(Class node) {
+    return node.enclosingLibrary.canonicalName.name == classLibraryUri &&
+    node.name == className;
+  }
+
+  @override
+  void visitConstructorInvocation(ConstructorInvocation node) {
+    final Class parentClass = node.target.parent as Class;
+    if (_matches(parentClass)) {
+      nonConstantLocations.add(<String, dynamic>{
+        'file': node.location.file.toString(),
+        'line': node.location.line,
+        'column': node.location.column,
+      });
+    }
+    super.visitConstructorInvocation(node);
```

Comment:
@aam - this helps me find non-const instances.  But it also finds instances that I believe get tree-shaken.

Is this because they're not actually tree-shaken, or is there some other way that I can determine whether they've been tree shaken?  Ideally, I'd like to ignore them if they were tree shaken.

Reply:
Are you looking at the kernel file that was compiled in --aot/--tfa mode(https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/compile.dart#L322)? If so, all of the stuff has already been tree-shaken out of it.

Reply:
Even with aot/tfa options, I still get these in the kernel.

Reply:
Are you able to see generally things being tree-shaken by aot compiler dill? Are there particular variables that seem to stick? You can use `pkg/vm/bin/dump_kernel.dart` to dump dill file and see who references them in aot dill file.